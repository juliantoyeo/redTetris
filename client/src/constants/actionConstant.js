export const ACTIONS = {
	ALERT_POP: 'ALERT_POP'
}

export const PLAYER_ACTIONS = {
	CREATE_PLAYER: 'CREATE_PLAYER',
	UPDATE_PLAYER: 'UPDATE_PLAYER'
}

export const ROOM_ACTIONS = {
	ADD_ROOM: 'ADD_ROOM',
	DELETE_ROOM: 'DELETE_ROOM',
	UPDATE_ROOM: 'UPDATE_ROOM'
}