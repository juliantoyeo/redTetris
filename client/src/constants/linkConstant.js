export const LINKS = {
	HOME: '/',
	GAME: '/game'
}