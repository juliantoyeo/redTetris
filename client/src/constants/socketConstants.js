export const SOCKET_RES = {
	PLAYER_CREATED: 'PLAYER_CREATED',
	PLAYER_NAME_EXIST: 'PLAYER_NAME_EXIST',
	PLAYER_NOT_EXIST: 'PLAYER_NOT_EXIST',
	ROOM_CREATED: 'ROOM_CREATED',
	ROOM_NAME_EXIST: 'ROOM_NAME_EXIST',
	ROOM_DOESNT_EXIST: 'ROOM_DOESNT_EXIST',
	ROOM_UPDATED: 'ROOM_UPDATED',
	ROOM_DELETED: 'ROOM_DELETED'
}

export const SOCKET_EVENTS = {
	START_GAME: 'START_GAME',
	RESTART_GAME: 'RESTART_GAME',
	CREATE_PLAYER: 'CREATE_PLAYER',
	UPDATE_PLAYER: 'UPDATE_PLAYER',
	FECTH_ALL_ROOM: 'FECTH_ALL_ROOM',
	CREATE_ROOM: 'CREATE_ROOM',
	UPDATE_ROOM: 'UPDATE_ROOM',
	DELETE_ROOM: 'DELETE_ROOM',
	UPDATE_BOARD: 'UPDATE_BOARD',
	GET_NEW_STACK: 'GET_NEW_STACK',
	UPDATE_STACK_INDEX: 'UPDATE_STACK_INDEX',
	ADD_BLOCKING_ROW: 'ADD_BLOCKING_ROW',
	SET_PLAYER_GAME_OVER: 'SET_PLAYER_GAME_OVER',
	GAME_IS_OVER: 'GAME_IS_OVER'
}
