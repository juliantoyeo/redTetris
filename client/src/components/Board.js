import React from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';

import Cell from './Cell';
import { GAME_MODE } from '../constants/gameConstant';

const mainContainerStyle = (props) => {
	return ({
		margin: '0 auto',
		display: 'grid',
		gridTemplateRows: `repeat(${props.height}, calc(${props.cellSize}vw))`,
		gridTemplateColumns: `repeat(${props.width}, 1fr)`,
		gridGap: '1px',
		width: '100%',
		maxWidth: `${props.cellSize * props.width}vw`,
		background: 'black'
	});
}

const Board = (props) => {
	const { board, cellSize, numberOfPlayer, mini, gameMode } = props;
	const getHeight = () => {
		if (!mini)
			return (board.length);
		else {
			let rowCount = 0
			for (let i = 0; i < board.length; i++) {
				if (checkShouldPrintRow(board[i]))
					rowCount++;
			}
			return rowCount;
		}
	}

	const checkShouldPrintRow = (row) => {
		if (!mini)
			return true;
		for (let i = 0; i < row.length; i++) {
			if (row[i] !== '0')
				return true;
		}
	}

	return (
		<div style={mainContainerStyle({ width: board[0].length, height: getHeight(), cellSize: cellSize })}>
			{_.map(board, (row) => {
				if (checkShouldPrintRow(row)) {
					return (
						_.map(row, (cell, x) => {
							let pieceType = cell
							if (gameMode === GAME_MODE.NO_GHOST_PIECE && cell === 'X') {
								pieceType = '0'
							}
							return (
								<Cell key={x} type={pieceType} numberOfPlayer={numberOfPlayer} />
							);
						})
					);
				}
			})}
		</div>
	)
}

Board.propTypes = {
	board: PropTypes.array,
	cellSize: PropTypes.number,
	numberOfPlayer: PropTypes.number,
	mini: PropTypes.bool,
	gameMode: PropTypes.string
};

export default Board;