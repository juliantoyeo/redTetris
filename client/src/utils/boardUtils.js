import { BOARD_SIZE } from '../constants/gameConstant'
import _ from 'lodash'

export const createBoard = () => {
	return (
		Array.from(new Array(BOARD_SIZE.HEIGHT), () => 
			new Array(BOARD_SIZE.WIDTH).fill('0')
		)
		// TEST_BOARD.I_SPIN
	)
}

export const checkCollision = (piece, boardWithLandedPiece, { x: moveX, y: moveY }) => {
	const newBoard = _.cloneDeep(boardWithLandedPiece)
	// console.log("boardWithLandedPiece", boardWithLandedPiece)
	// console.log(piece)
	for (let y = 0; y < piece.shape.length; y += 1) {
		for (let x = 0; x < piece.shape[y].length; x += 1) {
			const moveToY = y + piece.pos.y + moveY
			const moveToX = x + piece.pos.x + moveX
			if (piece.shape[y][x] !== '0') {
				// if (!newBoard[moveToY] || !newBoard[moveToY][moveToX] || newBoard[moveToY][moveToX][1] !== 'clear') {
				if (!newBoard[moveToY] || !newBoard[moveToY][moveToX] || newBoard[moveToY][moveToX] !== '0') {
					// console.log(piece.shape)
					// console.log("collision happen on x : ", x, " y : ", y, " moveToX : ", moveToX, " moveToY : ", moveToY)
					// colPos = { x: x, y: y}
					// return { collided: true, colPos: { x: x, y: y}}
					return true
				}
			}
		}
	}
}

export const getLandPosition = (piece, boardWithLandedPiece) => {
	// console.log("boardWithLandedPiece", boardWithLandedPiece)
	// console.log(piece)
	// let moveY = BOARD_SIZE.HEIGHT - 1
	let moveY = piece.pos.y
	if (piece.pos.y <= 0)
		moveY = 0
	// console.log("checked moveY : ", moveY, " y : ", moveY - piece.pos.y)
	while (!checkCollision(piece, boardWithLandedPiece, { x: 0, y: moveY - piece.pos.y}))
	{
		
		// if (moveY === BOARD_SIZE.HEIGHT - 1)
		// 	return { x: piece.pos.x, y: moveY - 1}
		moveY++
		// console.log("checked moveY : ", moveY, " y : ", moveY - piece.pos.y)
	}
	moveY -= 1
	if (moveY < 0 && piece.type !== 'I') moveY = 0
	// const landPos = { x: piece.pos.x, y: moveY}
	// console.log("landPos : ", landPos)
	return { x: piece.pos.x, y: moveY}
}
