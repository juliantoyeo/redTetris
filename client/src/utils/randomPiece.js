import { PIECES } from '../constants/gameConstant';

export const randomPiece = () => {
	const types = 'IJLOSZT';
	const randType = types[Math.floor(Math.random() * types.length)];
	return [PIECES[randType].shape[0], randType];
}
