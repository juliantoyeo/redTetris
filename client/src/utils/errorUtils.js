import { SOCKET_RES } from '../constants/socketConstants'

export const errorAlert = (error) => {
	switch (error) {
		case SOCKET_RES.PLAYER_NAME_EXIST:
			alert('Name exist, please try a different name');
			break;
		case SOCKET_RES.ROOM_NAME_EXIST:
			alert('Room name exist, please try a different name');
			break;
		case SOCKET_RES.ROOM_DOESNT_EXIST:
			alert('Room does not exist, please try again');
			break;
		default:
			return;
	}
}