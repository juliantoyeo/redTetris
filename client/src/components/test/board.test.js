import React from 'react';
import { shallow } from 'enzyme';

import Board from '../Board';
import { createBoard } from '../../utils/boardUtils';
import { CELL_SIZE, PIECES } from '../../constants/gameConstant';

describe('Test Board component', () => {
	it('Should render properly', () => {
		const wrapper = shallow(
			<Board 
				board={createBoard()} 
				cellSize={CELL_SIZE} 
				numberOfPlayer={1}
				mini={false}
			/>
		);
		expect(wrapper).toMatchSnapshot();
	});
	it('Should render mini board', () => {
		const wrapper = shallow(
			<Board 
				board={PIECES['I'].shape} 
				cellSize={CELL_SIZE} 
				numberOfPlayer={1}
				mini={true}
			/>
		);
		expect(wrapper).toMatchSnapshot();
	});
});