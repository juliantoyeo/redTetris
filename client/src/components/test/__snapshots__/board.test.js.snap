// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Board component Should render mini board 1`] = `
<div
  style={
    Object {
      "background": "black",
      "display": "grid",
      "gridGap": "1px",
      "gridTemplateColumns": "repeat(4, 1fr)",
      "gridTemplateRows": "repeat(4, calc(1.5vw))",
      "margin": "0 auto",
      "maxWidth": "6vw",
      "width": "100%",
    }
  }
>
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "0",
        "0",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type={
      Array [
        "I",
        "I",
        "I",
        "I",
      ]
    }
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "0",
        "0",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "0",
        "0",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "0",
        "I",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "0",
        "I",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "0",
        "I",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "0",
        "I",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "0",
        "0",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "0",
        "0",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type={
      Array [
        "I",
        "I",
        "I",
        "I",
      ]
    }
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "0",
        "0",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "I",
        "0",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "I",
        "0",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "I",
        "0",
        "0",
      ]
    }
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type={
      Array [
        "0",
        "I",
        "0",
        "0",
      ]
    }
  />
</div>
`;

exports[`Test Board component Should render properly 1`] = `
<div
  style={
    Object {
      "background": "black",
      "display": "grid",
      "gridGap": "1px",
      "gridTemplateColumns": "repeat(10, 1fr)",
      "gridTemplateRows": "repeat(20, calc(1.5vw))",
      "margin": "0 auto",
      "maxWidth": "15vw",
      "width": "100%",
    }
  }
>
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="0"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="1"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="2"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="3"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="4"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="5"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="6"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="7"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="8"
    numberOfPlayer={1}
    type="0"
  />
  <Memo(Cell)
    key="9"
    numberOfPlayer={1}
    type="0"
  />
</div>
`;
